fedimint-core/src/module/mod.rs

```diff
@@ -263,6 +263,75 @@ where
     }
 }
 
+// TODO: II is ridicolous
+pub trait ICommonModuleGen: Debug {
+    fn decoder(&self) -> Decoder;
+
+    fn module_kind(&self) -> ModuleKind;
+
+    fn hash_client_module(&self, config: serde_json::Value) -> anyhow::Result<sha256::Hash>;
+
+    fn to_dyn_client(&self) -> DynClientModuleGen;
+}
+
+impl ICommonModuleGen for DynClientModuleGen {
+    fn decoder(&self) -> Decoder {
+        (**self).decoder()
+    }
+
+    fn module_kind(&self) -> ModuleKind {
+        (**self).module_kind()
+    }
+
+    fn hash_client_module(&self, config: serde_json::Value) -> anyhow::Result<sha256::Hash> {
+        (**self).hash_client_module(config)
+    }
+
+    fn to_dyn_client(&self) -> DynClientModuleGen {
+        self.clone()
+    }
+}
+
+impl ICommonModuleGen for DynServerModuleGen {
+    fn decoder(&self) -> Decoder {
+        (**self).decoder()
+    }
+
+    fn module_kind(&self) -> ModuleKind {
+        (**self).module_kind()
+    }
+
+    fn hash_client_module(&self, config: serde_json::Value) -> anyhow::Result<sha256::Hash> {
+        (**self).hash_client_module(config)
+    }
+
+    fn to_dyn_client(&self) -> DynClientModuleGen {
+        DynClientModuleGen(Arc::new(self.clone()) as Arc<_>)
+    }
+}
+
+impl IClientModuleGen for DynServerModuleGen {
+    fn decoder(&self) -> Decoder {
+        <Self as ICommonModuleGen>::decoder(self)
+    }
+
+    fn module_kind(&self) -> ModuleKind {
+        <Self as ICommonModuleGen>::module_kind(self)
+    }
+
+    fn hash_client_module(&self, config: serde_json::Value) -> anyhow::Result<sha256::Hash> {
+        <Self as ICommonModuleGen>::hash_client_module(self, config)
+    }
+}
+
+pub trait IClientModuleGen: Debug + MaybeSend + MaybeSync {
```

Comment:
Perhaps a naive question. I see `IClientModuleGen` and `ICommonModuleGen` share some function definitions? Can one trait derive from the other? Although I suppose it would be weird to do `ICommonModuleGen : IClientModuleGen`

Reply:
I forgot to polish this part, and I've renamed some things and added some comments to make it more clear. Thanks for pointing it out!